notification-hubs-sdk/src/main/java/com/microsoft/windowsazure/messaging/notificationhubs/UserIdVisitor.java

```diff
@@ -0,0 +1,42 @@
+package com.microsoft.windowsazure.messaging.notificationhubs;
+
+import android.content.Context;
+import android.content.SharedPreferences;
+import com.microsoft.windowsazure.messaging.R;
+
+/**
+ * Associates the User ID with any Installation that it visits.
+ */
+class UserIdVisitor implements InstallationVisitor {
+    private static final String PREFERENCE_KEY = "userId";
+    private final SharedPreferences mPreferences;
+
+    public UserIdVisitor(Context context) {
+        mPreferences = context.getSharedPreferences(context.getString(R.string.installation_enrichment_file_key), Context.MODE_PRIVATE);
+    }
+
+    /**
+     *
+     * @param subject The {@link Installation} to be updated to include an UserId.
+     */
+    @Override
+    public void visitInstallation(Installation subject) {
+        subject.setUserId(getUserId());
+    }
+
+    /**
+     * Fetches the current UserId.
+     * @return The unique id, null if it has not been set.
+     */
+    public String getUserId() {
+        return mPreferences.getString(PREFERENCE_KEY, null);
+    }
+
+    /**
+     * Updates the unique identifier that will be associated with future calls to `visitInstallation`.
+     * @param userId The unique identifier to apply.
+     */
+    public void setUserId(String userId) {
```

Comment:
Let's update this so that it only changes if `userId` is different than the current value, and the method returns a boolean telling the caller whether or not an update was actually made.